tembo-pgmq-python/tembo_pgmq_python/async_queue.py

```diff
@@ -0,0 +1,241 @@
+        """Send a message to a queue."""
+        async with self.pool.acquire() as conn:
+            result = await conn.fetchrow(
+                "SELECT * FROM pgmq.send($1, $2::jsonb, $3);", queue, dumps(message).decode("utf-8"), delay
+            )
+        return result[0]
+
+    async def send_batch(self, queue: str, messages: List[dict], delay: int = 0) -> List[int]:
+        """Send a batch of messages to a queue."""
+        jsonb_array = [dumps(message).decode("utf-8") for message in messages]
+
+        async with self.pool.acquire() as conn:
+            result = await conn.fetch(
+                "SELECT * FROM pgmq.send_batch($1, $2::jsonb[], $3);",
+                queue,
+                jsonb_array,
+                delay,
+            )
+        return [message[0] for message in result]
+
+    async def read(self, queue: str, vt: Optional[int] = None) -> Optional[Message]:
+        """Read a message from a queue."""
+        batch_size = 1
+        async with self.pool.acquire() as conn:
+            rows = await conn.fetch("SELECT * FROM pgmq.read($1, $2, $3);", queue, vt or self.vt, batch_size)
+        messages = [
+            Message(msg_id=row[0], read_ct=row[1], enqueued_at=row[2], vt=row[3], message=loads(row[4])) for row in rows
+        ]
+        return messages[0] if len(messages) == 1 else None
+
+    async def read_batch(self, queue: str, vt: Optional[int] = None, batch_size=1) -> Optional[List[Message]]:
+        """Read a batch of messages from a queue."""
+        async with self.pool.acquire() as conn:
+            rows = await conn.fetch("SELECT * FROM pgmq.read($1, $2, $3);", queue, vt or self.vt, batch_size)
+
+        return [
+            Message(msg_id=row[0], read_ct=row[1], enqueued_at=row[2], vt=row[3], message=loads(row[4])) for row in rows
+        ]
+
+    async def read_with_poll(
+        self,
+        queue: str,
+        vt: Optional[int] = None,
+        qty: int = 1,
+        max_poll_seconds: int = 5,
+        poll_interval_ms: int = 100,
+    ) -> Optional[List[Message]]:
+        """Read messages from a queue with polling."""
+        async with self.pool.acquire() as conn:
+            rows = await conn.fetch(
+                "SELECT * FROM pgmq.read_with_poll($1, $2, $3, $4, $5);",
+                queue,
+                vt or self.vt,
+                qty,
+                max_poll_seconds,
+                poll_interval_ms,
+            )
+
+        return [
+            Message(msg_id=row[0], read_ct=row[1], enqueued_at=row[2], vt=row[3], message=loads(row[4])) for row in rows
+        ]
+
+    async def pop(self, queue: str) -> Message:
+        """Pop a message from a queue."""
+        async with self.pool.acquire() as conn:
+            rows = await conn.fetch("SELECT * FROM pgmq.pop($1);", queue)
+
+        messages = [
+            Message(msg_id=row[0], read_ct=row[1], enqueued_at=row[2], vt=row[3], message=loads(row[4])) for row in rows
```

Comment:
> loads(row[4])

I have not had a chance to dig into asyncpg a ton yet. Does it read the Postgres jsonb columns as a string, thus requiring us to deserialize into an object here?

Reply:
It's an excellent question.
Yes, by default. You can find it [here](https://magicstack.github.io/asyncpg/current/usage.html#type-conversion) by searching "json."

It's possible, though, to set up [per-connection type conversion](https://magicstack.github.io/asyncpg/current/usage.html#example-automatic-json-conversion).
I was thinking about implementing it, but it turned out there are not many places in the code where this conversion happens.

If you think it's crucial, I can implement it.

Reply:
It's not crucial. It just stood out to me because we don't have to explicity convert jsonb to dict when using psycopg. Thanks for the link to docs.